projects/gltf_basic_materials/GltfBasicMaterials.cpp

```diff
@@ -0,0 +1,339 @@
+            gpCreateInfo.PS                                 = {mPS.Get(), "psmain"};
+            gpCreateInfo.topology                           = grfx::PRIMITIVE_TOPOLOGY_TRIANGLE_LIST;
+            gpCreateInfo.polygonMode                        = grfx::POLYGON_MODE_FILL;
+            gpCreateInfo.cullMode                           = grfx::CULL_MODE_BACK;
+            gpCreateInfo.frontFace                          = grfx::FRONT_FACE_CCW;
+            gpCreateInfo.depthReadEnable                    = true;
+            gpCreateInfo.depthWriteEnable                   = true;
+            gpCreateInfo.blendModes[0]                      = grfx::BLEND_MODE_NONE;
+            gpCreateInfo.outputState.renderTargetCount      = 1;
+            gpCreateInfo.outputState.renderTargetFormats[0] = GetSwapchain()->GetColorFormat();
+            gpCreateInfo.outputState.depthStencilFormat     = GetSwapchain()->GetDepthFormat();
+            gpCreateInfo.pPipelineInterface                 = mPipelineInterface;
+
+            gpCreateInfo.vertexInputState.bindingCount = CountU32(vertexBindings);
+            for (uint32_t i = 0; i < gpCreateInfo.vertexInputState.bindingCount; ++i) {
+                gpCreateInfo.vertexInputState.bindings[i] = vertexBindings[i];
+            }
+
+            PPX_CHECKED_CALL(GetDevice()->CreateGraphicsPipeline(&gpCreateInfo, ppPipeline));
+        };
+
+        // Pipelines
+        CreatePipeline("MaterialVertex.vs", "StandardMaterial.ps", &mStandardMaterialPipeline);
+        CreatePipeline("MaterialVertex.vs", "UnlitMaterial.ps", &mUnlitMaterialPipeline);
+        CreatePipeline("MaterialVertex.vs", "ErrorMaterial.ps", &mErrorMaterialPipeline);
+
+        // Compile pipelines for mmaterials
+        for (auto it : mMaterialIndexMap) {
+            auto pMaterial = it.first;
+            auto ident     = pMaterial->GetIdentString();
+
+            if (ident == PPX_MATERIAL_IDENT_STANDARD) {
+                mMaterialPipelineMap[pMaterial] = mStandardMaterialPipeline;
+            }
+            else if (ident == PPX_MATERIAL_IDENT_UNLIT) {
+                mMaterialPipelineMap[pMaterial] = mUnlitMaterialPipeline;
+            }
+            else {
+                mMaterialPipelineMap[pMaterial] = mErrorMaterialPipeline;
+            }
+        }
+    }
+}
+
+void GltfBasicMaterialsApp::Shutdown()
+{
+    delete mScene;
+    delete mPipelineArgs;
+}
+
+void GltfBasicMaterialsApp::Render()
+{
+    PerFrame& frame = mPerFrame[0];
+
+    grfx::SwapchainPtr swapchain = GetSwapchain();
+
+    uint32_t imageIndex = UINT32_MAX;
+    PPX_CHECKED_CALL(swapchain->AcquireNextImage(UINT64_MAX, frame.imageAcquiredSemaphore, frame.imageAcquiredFence, &imageIndex));
+
+    // Wait for and reset image acquired fence
+    PPX_CHECKED_CALL(frame.imageAcquiredFence->WaitAndReset());
+
+    // Wait for and reset render complete fence
+    PPX_CHECKED_CALL(frame.renderCompleteFence->WaitAndReset());
```

Comment:
Wait on renderCompleteFence before AcquireNextImage, see #445

Reply:
Question: Is the wait for imageAcquiredFence() required here?
Could we do:
- renderCompleteFence.WaitAndReset()
- acquireNextImage(imageAcquiredSemaphore)
- graphics.Submit(wait=imageAcquiredSemaphore, signal=renderCompleteSemaphore + fence)
- present(wait=renderCompleteSemaphore)

Reply:
@Keenuts There was a thread on this in our chat last November, where someone mentioned acquireNextImage return with invalid index, I am not sure how though.